packages/ui/src/Modal.tsx

```diff
@@ -0,0 +1,166 @@
+ * both part of the union separately, as picking the result discriminates the type and breaks the constrains of A | B.
+ *
+ * When Storybook supports tsc@4.1, this can be replaced with:
+ * type MapObjectUnion<T> = T extends Record<string, unknown> ? ModalActionMappedProps<T> : T
+ * type ModalActionMappedProps<T> = { [U in keyof T as `action${Capitalize<U>}`]: T[U] }
+ */
+type ModalActionDisabledProps =
+  | {
+      actionDisabledTooltip: ButtonDisabledProps['disabledTooltip']
+      actionDisabled: ButtonDisabledProps['disabled']
+    }
+  | {
+      actionDisabledTooltip?: never
+      actionDisabled?: never
+    }
+
+type ModalActionIconProps =
+  | {
+      actionIconLeft: ButtonAccessibleIconLeftProps['iconLeft']
+      actionIconLeftAriaLabel: ButtonAccessibleIconLeftProps['iconLeftAriaLabel']
+    }
+  | {
+      actionIconLeft?: never
+      actionIconLeftAriaLabel?: never
+    }
+
+type ModalActionCoreProps =
+  | {
+      // Note: Source of `onClick` is nullable property from ButtonHTMLAttributes<HTMLButtonElement>
+      onAction: NonNullable<ButtonProps['onClick']>
+      actionLabel: string
+    }
+  | { onAction?: never; actionLabel?: never }
+
+export type ModalActionProps = ModalActionCoreProps & {
+  actionKind?: ButtonProps['kind']
+} & ModalActionDisabledProps &
+  ModalActionIconProps
+
+export type ModalProps = {
+  closable?: boolean
+  children?: ReactNode
+  title: string
+  onClose?: ReactModalProps['onRequestClose']
+  footerClassName?: string
+} & DataTestProp &
+  ModalActionProps &
+  Exclude<ReactModalProps, 'onRequestClose' | 'shouldFocusAfterRender' | 'shouldReturnFocusAfterClose'>
+
+/*
+ * ### Purpose
+ * Occasionally there's a user-story that's not a part of the main user flow. Such action can be contained in a Modal.
+ * Modals can contain components like forms, menus, tables etc.
+ * Usually modals are used to drive a complex action. In case there is a simple action (e.g. confirmation), consider using a Dialog.
+ *
+ * ### General Info
+ * - Modal always contains a title, icon "X" in the header, content and active buttons in the footer.
+ * - No interactions on the underlying page can be performed while a Modal.
+ * - Content beneath the modal is covered by a "blanket".
+ * - To close the Modal, use "Cancel" button in the footer, "X" button in the header, press "Esc" key or click anywhere on the "blanket".
+ */
+export const Modal: FC<ModalProps> = ({
+  'data-test': dataTest,
+  children,
+  className,
+  footerClassName,
+  closable = true,
+  onClose,
+  overlayClassName,
+  title,
+  ...restWithActionButton
+}) => {
+  const {
+    actionKind,
+    actionLabel,
+    // Disabled
+    actionDisabled = false,
+    actionDisabledTooltip,
+    // Icon
+    actionIconLeft,
+    actionIconLeftAriaLabel,
+    // Rest
+    onAction,
+    ...rest
+  } = restWithActionButton
+
+  return (
+    <ReactModal
+      contentLabel={title}
+      className={cn(styles.modal, className)}
```

Comment:
@jankoritak should we also use `portalClassName` to set dimensions for the portal? Currently, it is unstyled. Because of it, its width is 0px and the overlay cannot stretch.
![image](https://user-images.githubusercontent.com/12794628/104133517-258a8680-5395-11eb-8979-655c223598ab.png)

Reply:
I must have broke it during cleanup. 🤦

Reply:
Fixed.